File: Algorithm/Sort/실패율.py
# 실패율
# https://programmers.co.kr/learn/courses/30/lessons/42889


def solution(N, stages):
    fail_rate = []
    length = len(stages)
    for i in range(1, N+1):
        if length == 0:
            for j in range(i, N+1):
                fail_rate.append((0, j))
            break
        fail_count = stages.count(i)
        fail_rate.append((-1*fail_count/length, i))
        length -= fail_count
    fail_rate.sort()
    return list(map(lambda x: x[1], fail_rate))


# 처음 풀이
# 굳이 filter와 sum으로 비슷한 연산을 반복해서 시간적으로 비효울적이었음
# def solution(N, stages):
#     failRate = []
#     for i in range(1, N+1):
#         stages = list(filter(lambda x: x >= i, stages))
#         if len(stages):
#             failRate.append((0, i))
#             continue
#         failCount = sum(stage == i for stage in stages)
#         clearCount = len(stages) - failCount
#         failRate.append((-1*failCount/(failCount + clearCount), i))
#     failRate.sort()
#     return list(map(lambda x: x[1], failRate))

File: Algorithm/Sort/실패율.js
function solution(N, stages) {
    const failRates = [];
    
    stages.sort((a, b) => a - b);
    let length = stages.length;
    let index = 0; // stages 탐색할 index
    
    // 스테이지 별로 실패율 계산
    for (let stage = 1; stage <= N; stage++) {
        if (length === 0 || index >= stages.length) {
            failRates.push([0, stage]);
            continue;
        }
        
        let failCount = 0;
        while (stages[index] === stage)  {
            index++;
            failCount++;
        }
       
        failRates.push([failCount / length, stage]);
        length -= failCount;
    }
 
    return failRates.sort((a, b) => {
        if (a[0] - b[0] !== 0)
            return b[0] - a[0];
        return a[1] - b[1];
    }).map((x) => x[1]);;
}